### PR TITLE
changefeedccl: correctly handle old-style protobufs in rowfetcher_cache

### LINKED ISSUE #82309: cdc: changefeeds stop emitting events after an upgrade
**Describe the problem**

Changefeeds silently fail to emit data if they were created before 22.1 and are now running on 22.1.

**To Reproduce**

Create a changefeed using 21.2 with resolved timestamps. Pause it. Upgrade to 22.1. Resume the existing changefeed and write to the table. You will see resolved timestamps in your sink but not data.

### PR DESCRIPTION
**Pull request by HonoreDB**
Fixes https://github.com/cockroachdb/cockroach/issues/82309.

Release note (bug fix): Fixed a bug where changefeeds created before upgrading to 22.1 would silently fail to emit any data other than resolved timestamps.

### CODE REVIEW

**Review by miretskiy — CHANGES_REQUESTED** [review]
> Reviewed 3 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82311)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @gh-casper and @HonoreDB)

___
*[`pkg/ccl/changefeedccl/cdcevent/event.go` line 357 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82311#-N3W9BFw-kALWczeFew9:-N3W9BFw-kALWczeFewA:ber3o9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/81ac4829a644e39ee6fe96c5cd343edc193d6bb2/pkg/ccl/changefeedccl/cdcevent/event.go#L357)):*
> ```Go
> 		cfg.CollectionFactory,
> 		cfg.DB,
> 		changefeedbase.AllTargets(details),
> ```

Let's be extra paranoid and add a check in newRowFetcherCache to require that at least 1 specification provided.  Probably need to change this method to return error...
___
*[`pkg/ccl/changefeedccl/changefeedbase/validate.go` line 114 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82311#-N3W91aN2Phri0hLEIHw:-N3W91aN2Phri0hLEIHx:b-gf60mp) ([raw file](https://github.com/cockroachdb/cockroach/blob/81ac4829a644e39ee6fe96c5cd343edc193d6bb2/pkg/ccl/changefeedccl/changefeedbase/validate.go#L114)):*
> ```Go
> 	return warnings
> }
> ```

public exported methods need a comment.
___
*[`pkg/ccl/changefeedccl/cdcevent/event_test.go` line 394 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82311#-N3W9kaD54wWZqv15An0:-N3W9kaD54wWZqv15An1:b-ci99or) ([raw file](https://github.com/cockroachdb/cockroach/blob/81ac4829a644e39ee6fe96c5cd343edc193d6bb2/pkg/ccl/changefeedccl/cdcevent/event_test.go#L394)):*
> ```Go
> 		actions           []string
> 		expectMainFamily  []decodeExpectation
> 		expectOnlyCFamily []decodeExpectation
> ```

you probably don't need to have these extra family expectations -- old style feeds didn't support that.
___
*[`pkg/ccl/changefeedccl/cdcevent/event_test.go` line 406 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82311#-N3WA4IR4ecSBDiOAGkg:-N3WA4IR4ecSBDiOAGkh:b-z9ua3j) ([raw file](https://github.com/cockroachdb/cockroach/blob/81ac4829a644e39ee6fe96c5cd343edc193d6bb2/pkg/ccl/changefeedccl/cdcevent/event_test.go#L406)):*
> ```Go
> 				},
> 			},
> 		},
> ```

you might not need to have test case harness like this .  If it's just 1 test case, just run it.
Or... Add few more test cases.
Perhaps with deletions too.

**Comment by HonoreDB** [comment]
Comments addressed.

File: pkg/ccl/changefeedccl/changefeed_processors.go [hunk 1]
```diff
@@ -297,9 +297,16 @@ func (ca *changeAggregator) Start(ctx context.Context) {
 		return
 	}
 
-	ca.eventConsumer = newKVEventToRowConsumer(
+	ca.eventConsumer, err = newKVEventToRowConsumer(
 		ctx, ca.flowCtx.Cfg, ca.frontier.SpanFrontier(), kvFeedHighWater,
 		ca.sink, ca.encoder, ca.spec.Feed, ca.knobs, ca.topicNamer)
+
+	if err != nil {
+		// Early abort in the case that there is an error creating the sink.
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/changefeed_processors.go:305]: Comment needs to be updated a bot

**Review by miretskiy — APPROVED** [review]

File: pkg/ccl/changefeedccl/changefeed_processors.go (hunk 1 above)
Comment (HonoreDB) [reply pkg/ccl/changefeedccl/changefeed_processors.go]: Updated.

**Comment by HonoreDB** [comment]
bors r=[miretskiy]

Thanks for the find and the quick reviews!

**Comment by HonoreDB** [comment]
bors cancel

File: pkg/ccl/changefeedccl/event_processing.go [hunk 2]
```diff
@@ -75,7 +79,7 @@ func (c *kvEventToRowConsumer) topicForEvent(eventMeta cdcevent.Metadata) (Topic
 			return topic, nil
 		}
 	}
-	for _, s := range c.details.TargetSpecifications {
+	for _, s := range AllTargets(c.details) {
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/event_processing.go:82]: Any chance to add a test for this? 
A direct call to topicForEvent with 2 protos -- 1 old 1 new.
You don't even need to construct full "consumer"-- just initialize "details" field.

**Review by miretskiy — CHANGES_REQUESTED** [review]

File: pkg/ccl/changefeedccl/event_processing.go (hunk 2 above)
Comment (HonoreDB) [reply pkg/ccl/changefeedccl/event_processing.go:82]: I made an attempt (code here: https://gist.github.com/HonoreDB/1386d17dedbd4150702f29495a2a6b1f) but am missing some kind of cleanup step and am probably not going to have time to figure it out tonight, have to wake up at 5 AM tomorrow. 
```
panic: rangefeed-system-monitor: unexpected 20480 leftover bytes

goroutine 940 [running]:
github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic({0x9e77748, 0xc00241ffb0}, 0xc00120e000, {0x88753dc, 0xc002c24800}, {0xc001df7730, 0x47c7c40, 0xc001df7718})
	github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:376 +0x1af
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0xc000ada600, {0x9e77748, 0xc00241ffb0}, 0x1)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:426 +0x22a
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(...)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:406
github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*BudgetFactory).Stop(0xc00125c480, {0x9e77748, 0xc00241ffb0})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/budget.go:356 +0x45
```

**Comment by miretskiy** [comment]
Feel free to leave a comment re additional test.  Giving 👍 since its critical this goes in.

**Comment by shermanCRL** [comment]
@stevendanna mentions a follow-on, consider having `AllTargets` return distinct type.

**Comment by miretskiy** [comment]
Filed https://github.com/cockroachdb/cockroach/issues/82342

**Comment by miretskiy** [comment]
Subsumed by https://github.com/cockroachdb/cockroach/pull/82371

**Pull request closed without merging**